Libs/MRML/DisplayableManager/vtkMRMLSliceIntersectionInteractionRepresentation.cxx

```diff
@@ -0,0 +1,2333 @@
+#include "vtkMatrix4x4.h"
+#include "vtkObjectFactory.h"
+#include "vtkPoints.h"
+#include "vtkPolyDataAlgorithm.h"
+#include "vtkPolyDataMapper2D.h"
+#include "vtkProperty2D.h"
+#include "vtkPlane.h"
+#include "vtkRenderer.h"
+#include "vtkRenderWindow.h"
+#include "vtkSphereSource.h"
+#include "vtkTransform.h"
+#include "vtkTransformPolyDataFilter.h"
+#include "vtkTextMapper.h"
+#include "vtkTextProperty.h"
+#include "vtkTubeFilter.h"
+#include "vtkWindow.h"
+
+// MRML includes
+#include <vtkMRMLInteractionEventData.h>
+
+// Visualization modes
+enum
+  {
+  ShowIntersection = 0,
+  HideIntersection = 1,
+  };
+
+// Handles type
+enum
+  {
+  Arrows = 0,
+  Circles = 1,
+  };
+
+// Settings
+static const double VISUALIZATION_MODE = HideIntersection;
+static const double HANDLES_TYPE = Arrows;
+static const bool HANDLES_ALWAYS_VISIBLE = false;
+static const double OPACITY_RANGE = 1000.0;
+static const double FOV_HANDLES_MARGIN = 0.03; // 3% margin
+static const double HIDE_INTERSECTION_GAP_SIZE = 0.05; // 5.0% of the slice view width
+static const double INTERACTION_SIZE_PIXELS = 20.0;
+static const double HANDLES_MIN_LINE_LENGTH = 50.0;
+
+// Intersection line
+static const double INTERSECTION_LINE_RESOLUTION = 50; // default = 8
+static const double INTERSECTION_LINE_EXTRA_THICKNESS = 1.0; // extra thickness with respect to normal slice intersection display
+
+// Handles
+static const double HANDLES_CIRCLE_THETA_RESOLUTION = 100; // default = 8
+static const double HANDLES_CIRCLE_PHI_RESOLUTION = 100; // default = 8
+static const double SLICEOFSSET_HANDLE_DEFAULT_POSITION[3] = { 0.0,0.0,0.0 };
+static const double SLICEOFSSET_HANDLE_DEFAULT_ORIENTATION[3] = { 0.0,1.0,0.0 };
+static const double SLICEOFSSET_HANDLE_CIRCLE_RADIUS = 7.0;
+static const double SLICEOFSSET_HANDLE_ARROW_RADIUS = 3.0;
+static const double SLICEOFSSET_HANDLE_ARROW_LENGTH = 60.0;
+static const double SLICEOFSSET_HANDLE_ARROW_TIP_ANGLE = 27; // degrees
+static const double ROTATION_HANDLE_DEFAULT_POSITION[3] = { 0.0,0.0,0.0 };
+static const double ROTATION_HANDLE_DEFAULT_ORIENTATION[3] = { 0.0,1.0,0.0 };
+static const double ROTATION_HANDLE_CIRCLE_RADIUS = 10.0;
+static const double ROTATION_HANDLE_ARROW_RADIUS = 3.0;
+static const double ROTATION_HANDLE_ARROW_LENGTH = 60.0;
+static const double ROTATION_HANDLE_ARROW_TIP_ANGLE = 27; // degrees
+static const double TRANSLATION_HANDLE_OUTER_RADIUS = 9.0;
+static const double TRANSLATION_HANDLE_INNER_RADIUS = 7.0;
+
+vtkStandardNewMacro(vtkMRMLSliceIntersectionInteractionRepresentation);
+vtkCxxSetObjectMacro(vtkMRMLSliceIntersectionInteractionRepresentation, MRMLApplicationLogic, vtkMRMLApplicationLogic);
+
+class SliceIntersectionInteractionDisplayPipeline
+{
+  public:
+
+    //----------------------------------------------------------------------
+    SliceIntersectionInteractionDisplayPipeline()
+    {
+      // Intersection line 1
+      this->IntersectionLine1 = vtkSmartPointer<vtkLineSource>::New();
```

Comment:
This sounds like a lot of work and we cannot spend more weeks on this project, it has already been several months. Could we add an issue for this one?

Reply:
All these hardcodings for two slices are bad - they make the code longer and less clean, and results for all the layouts that have more than 3 slices in a view group are bad. The computations are incorrect for two slices, too, if they are not orthogonal (see video attached to one of the comments). I can understand that a lot of things had to be learned for developing this and it took a long time, but I don't think this could go in as is. At some point @Sunderlandkyl or I can take over some tasks, but preferably the complexity of the current code should be made a bit more manageable before that.

We have the correct rotation formula for N slices (for the Ctrl+Alt+Click-and-drag rotation). Also, I expect that logic to handle N slices instead of 2 will make the code shorter and more clear (copy-paste will be replaced by a loop at most places).

Reply:
Thanks for the comment @lassoan . Actually, I am not hardcoding the number of intersection lines in the slice view to 2 as you indicate. Each intersecting slice node is managed by a different display pipeline, so any number of slice views and intersections should be compatible with the current infrastructure. Although some fixes may be needed to adapt it to different use cases, I believe the infrastructure is the proper one.

I create 2 line objects (vtkLineSource) for each display pipeline to be able to render two separate segments for each intersection line. This is needed to display the intersection lines with a gap around the slice intersection point. See image below:

![image](https://user-images.githubusercontent.com/10816661/141780257-4d83127d-c5e4-4a29-a7e5-09423a1beeb2.png)

That's why two line objects are created.

Reply:
That's good then! I was puzzled by those many duplication of sources, actors, mappers, properties and their settings and updates throughout the code. There is room for improvement here, as there is lots of repetitive code, but as long as things work and performance is acceptable, it is not required to change this aspect of the code at this time.

Reply:
One more thing related to this: In general, slices don't intersect each other in a single point. Instead, you get a set of intersection points. You can use the average of these intersection points as center of rotation. Probably the intersection lines could be blanked out within a certain distance from this single center of rotation point, so you would still need to have only one gap in each intersection line.